juju/juju
# hook tools for state cache get/delete/set

### Checklist

        We have this one the list of work to do.

----

## Description of change

Add hook tools for state-cache-get, state-cache-set and state-cache-delete.  This is part of the uniter improvements: server side state storage for charms.

## QA steps
Bootstrap and deploy a unit to use the hook tools on, verify the cache is empty to start.
```sh
juju bootstrap localhost
juju deploy ubuntu
juju run --unit ubuntu/0 state-get
```
Set a key value pair and verify.
```sh
juju run --unit ubuntu/0 state-set one=two
juju run --unit ubuntu/0 state-get
```
Set key value pair with set to empty string and verify.
```sh
juju run --unit ubuntu/0 state-set one= two=three
juju run --unit ubuntu/0 state-get
```
Remove a key value pair with delete and verify.
```sh
juju run --unit ubuntu/0 state-set one=two
juju run --unit ubuntu/0 state-delete one
juju run --unit ubuntu/0 state-get
```
Add multiple key value pairs at one
```sh
juju run --unit ubuntu/0 state-set one=two three=four five=six
juju run --unit ubuntu/0 state-get
```

Use state-set with the --file option.  Add file to /var/lib/juju/agents/unit-ubuntu-0/charm/ to be used, with yaml contents.

## Documentation changes

Will be done as part of juju 2.8
